cro-clib/src/transaction.rs

```diff
@@ -0,0 +1,318 @@
+use crate::types::get_string;
+use crate::types::CroStakedState;
+use crate::types::{CroAddress, CroAddressPtr, CroResult};
+use chain_core::init::coin::Coin;
+pub use chain_core::init::network::Network;
+use chain_core::state::account::{
+    StakedState, StakedStateAddress, StakedStateOpAttributes, StakedStateOpWitness, UnbondTx,
+    WithdrawUnbondedTx,
+};
+use chain_core::tx::data::access::{TxAccess, TxAccessPolicy};
+use chain_core::tx::data::address::ExtendedAddr;
+use chain_core::tx::data::attribute::TxAttributes;
+use chain_core::tx::data::output::TxOut;
+use chain_core::tx::TransactionId;
+use chain_core::tx::TxAux;
+use client_common::tendermint::types::AbciQueryExt;
+use client_common::tendermint::{Client, WebsocketRpcClient};
+use client_common::{ErrorKind, Result, ResultExt};
+use client_common::{PublicKey, SignedTransaction};
+use client_core::cipher::DefaultTransactionObfuscation;
+use client_core::TransactionObfuscation;
+use parity_scale_codec::Decode;
+use parity_scale_codec::Encode;
+use std::collections::BTreeSet;
+use std::os::raw::c_char;
+use std::ptr;
+use std::str::FromStr;
+use std::string::ToString;
+fn do_cro_unbond(
+    network: u8,
+    nonce: u64,
+    from_address: &CroAddress,
+    to_address_user: &str,
+    amount: u64,
+) -> Result<Vec<u8>> {
+    let to_address = StakedStateAddress::from_str(&to_address_user).chain(|| {
+        (
+            ErrorKind::DeserializationError,
+            format!("Unable to deserialize to_address ({})", to_address_user),
+        )
+    })?;
+    let value =
+        Coin::new(amount).chain(|| (ErrorKind::DeserializationError, "Invalid Coin Amount"))?; // carson unit
+    let attributes = StakedStateOpAttributes::new(network);
+    let transaction: UnbondTx = UnbondTx::new(to_address, nonce, value, attributes);
+    let from_private = &from_address.privatekey;
+    let signature: StakedStateOpWitness = from_private
+        .sign(transaction.id())
+        .map(StakedStateOpWitness::new)?;
+    let result = TxAux::UnbondStakeTx(transaction, signature);
+    let encoded = result.encode();
+    Ok(encoded)
+}
+
+/// staked -> staked
+/// network: networkid
+/// nonce: nonce of the staked state, use cro_get_staked_state to get this nonce
+/// from_ptr: staking address
+/// to_address_user:staking address
+/// amount: carson unit   1 carson= 0.0000_0001 cro
+/// output: signed tx encoded
+#[no_mangle]
+/// # Safety
```

Comment:
https://doc.rust-lang.org/book/ch14-02-publishing-to-crates-io.html#making-useful-documentation-comments
> Safety: If the function is unsafe to call (we discuss unsafety in Chapter 19), there should be a section explaining why the function is unsafe and covering the invariants that the function expects callers to uphold.

I think rather than having these empty sections, it'd be good to explain what is expected -- I guess it's expecting that pointers point to valid / previously created data with `cro_create_*`? No null pointers? That string is null-terminated?

Reply:
ok